Add public API for connecting EATT channels

hermabe opened the pull request:
This PR adds a public API for EATT connection.
- `bt_eatt_connect()` is made public.
- The application can register a callback to get notified when an EATT channel is connected.
- If `CONFIG_BT_EATT_AUTO_CONNECT` is set, the central will automatically try to connect `CONFIG_BT_EATT_MAX` EATT channels when a connection to a peripheral is established.

The EATT connection test from the L2CAP collision mitigation PR is updated to use the new API and test the auto-connection feature.

hermabe committed:
Bluetooth: Host: Make bt_eatt_connect connect up to BT_EATT_MAX channels

bt_eatt_connect may now send multiple credit based connection requests.
Previously it would only send one and thus only allow 5 EATT channels to
be connected.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

hermabe committed:
Bluetooth: Host: Automatically create EATT channels on connection

If the Kconfig BT_EATT_AUTO_CONNECT is enabled (on by default), the
host will try to connect BT_EATT_MAX EATT channels when a connection
to a peer is established.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

hermabe committed:
Bluetooth: Host: Cancel EATT connection work in att_reset

When att->conn is set to NULL in att_reset, EATT connection will fail.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

hermabe commented on include/bluetooth/att.h:
```diff
@@ -59,12 +59,53 @@ extern "C" {
 
 #endif /* CONFIG_BT_TESTING */
+
+/** @brief Connect Enhanced ATT channels
```
Agree that "bearer" is better, but I changed it to "channel" here and in a previous PR to be consistent with what is already in  use.

Thalley replied:
`bearer` is the core term, but besides that `channel` is a more fitting term to what they actually are, as well as being consistent with Zephyr naming. 

Not sure what the right course of action is here.

hermabe replied:
If we change it to "bearer", then we should do that in the whole host. Shall we keep it as "channel" here and have a separate PR to change the names both in code and in Kconfig to either "bearer" or "channel"?

Thalley replied:
By "whole host", do you mean "just" att.c/h? Channel is the right (core) term for e.g. L2CAP, and somewhat the right term for ISO as well.

Thalley replied:
Btw, I suggest to use "channel" to be consistent with the code base for now. If we want to change it to bearer, it'll be a larger change that will also affect the ATT API I reckon.

hermabe replied:
I meant that we would have to change it everywhere, but I guess it is just ATT. `bt_l2cap_chan` and `bt_l2cap_l2_chan` are separate from this are correctly named "channel".

hermabe commented on subsys/bluetooth/host/att.c:
```diff
@@ -3088,7 +3106,16 @@ int bt_eatt_connect(struct bt_conn *conn, uint8_t num_channels)
 
-	return bt_l2cap_ecred_chan_connect(conn, chan, BT_EATT_PSM);
+	while (created < i) {
+		err = bt_l2cap_ecred_chan_connect(conn, &chan[created], BT_EATT_PSM);
```
Yes, can comment this better. `bt_l2cap_ecred_chan_connect` takes an array with up to 5 elements and must be null-terminated if shorter.

Thalley commented on include/bluetooth/att.h:
```diff
@@ -59,12 +59,53 @@ extern "C" {
+ * @param conn The connection to send the request on
+ * @param num_channels The number of Enhanced ATT beaeres to request.
+ *
+ * @return 0 in case of success or negative value in case of error.
```
There's also the use of `@retval`, e.g. `@retval -EINVAL if invalid parameters supplied` that I've grown to be a fan of.

Thalley commented on include/bluetooth/att.h:
```diff
@@ -59,12 +59,53 @@ extern "C" {
 
+/* ATT channel information */
+struct bt_att_chan_info {
+	/* The connection the ATT channel belongs to */
```
Would it not be better to add borrowed reference only for those where the application needs to do `bt_conn_unref()`?

hermabe replied:
Bump @alwa-nordic

alwa-nordic replied:
> Would it not be better to add borrowed reference only for those where the application needs to do bt_conn_unref()?

I agree.

Thalley commented on include/bluetooth/att.h:
```diff
@@ -59,12 +59,53 @@ extern "C" {
+ * may be created.
+ *
+ * @param conn The connection to send the request on
+ * @param num_channels The number of Enhanced ATT beaeres to request.
```
Is there any value range here? Can the application e.g. request 125 channels? 
I suppose we also have a local max number of channels?

hermabe replied:
I can add a reference to `CONFIG_BT_EATT_MAX`.

hermabe replied:
I can add a reference to `CONFIG_BT_EATT_MAX`.

hermabe replied:
I can add a reference to `CONFIG_BT_EATT_MAX`.

Thalley replied:
Please do :)

Thalley commented on include/bluetooth/att.h:
```diff
@@ -59,12 +59,53 @@ extern "C" {
+ *
+ *  Register callbacks to monitor the state of ATT.
+ *
+ *  @param cb Callback struct.
```
What do you mean by `reuse the referenced object`?

alwa-nordic replied:
> What do you mean by reuse the referenced object?

I meant don't call `bt_eatt_cb_register` twice with the same object. (I guess it's not really a concern in this case. But it is in eg. `bt_gatt_notify`.)

Thalley commented on subsys/bluetooth/host/Kconfig.gatt:
```diff
@@ -43,6 +43,14 @@ config BT_EATT_MAX
 
+
+config BT_EATT_AUTO_CONNECT
+	bool "Automatically connect EATT bearers when a link is established"
```
I think we need to decide between EATT bearers or EATT channels, and use the same term everywhere.

Vudentz replied:
That is because the spec doesn't differentiate an EATT bearer to an ATT bearer, ATT bearer are uniquely identified by its L2CAP CID and thus can save state (CCC states, etc), EATT bearer uses L2CAP Connection Oriented Channels and cannot save state thus why channels was used instead, though this explanation could probably be added as part of the documentation.

Thalley commented on subsys/bluetooth/host/Kconfig.gatt:
```diff
@@ -43,6 +43,14 @@ config BT_EATT_MAX
+	bool "Automatically connect EATT bearers when a link is established"
+	default y
+	help
+	  The central will try to connect BT_EATT_MAX enhanced ATT bearers when a
+	  connection to a peripheral is established.
```
So this is only for the central? 

If so, then we should add dependency on `BT_CENTRAL`. 
Is there any reason why the peripheral shouldn't attempt to create the EATT bearers/channels?

hermabe replied:
Chose that only the central initiates the request to avoid collisions. With two zephyr devices this is handled by the L2CAP collision mitigation procedure, but it may not work as well with other devices.

Bluez also chose to only auto-connect if it is the central.

Thalley replied:
The alternative is of course to have a delay for one of the sides (e.g. 2 x connection interval delay on the peripheral). 

IIRC there are no recommendations that the central or peripheral should attempt to create these on connection. In the case of LE Audio where we definitely want to use EATT channels, it may be useful for the peripheral side to also auto-initiate these in case we are connected to a central that doesn't connect them.

hermabe replied:
Yes, we can let the peripheral delay the connection request as is done in the collision mitigation procedure. This will work fine with two zephyr devices. Depending on what other stacks does we will have to rely on the collision mitigation working.

Thalley replied:
> Depending on what other stacks does we will have to rely on the collision mitigation working.

We will need to do that regardless in case that we connect to a non-Zephyr peripheral that auto-connects without delay as well :)

hermabe replied:
Changed it so both central and peripheral tries to connect. We may want to consider a backoff-mechanism in the future if the spec-required collision mitigation is insufficient.

Thalley commented on subsys/bluetooth/host/att.c:
```diff
@@ -136,6 +137,13 @@ void att_sent(struct bt_conn *conn, void *user_data)
 	}
 }
 
+static sys_slist_t callback_list;
```
To be fair, this is the name used in `conn.c` and `gatt.c`.

hermabe replied:
I just did as it is in GATT, but can rename it.

Thalley commented on subsys/bluetooth/host/att.c:
```diff
@@ -3067,8 +3082,11 @@ int bt_eatt_connect(struct bt_conn *conn, uint8_t num_channels)
-	struct bt_l2cap_chan *chan[CONFIG_BT_EATT_MAX] = {};
+	/* bt_l2cap_ecred_chan_connect expects a null-terminated array */
+	struct bt_l2cap_chan *chan[CONFIG_BT_EATT_MAX + 1] = {};
+	int created = 0;
```
Suggest to make `created` and `i` unsigned values (e.g. `size_t`)

Thalley commented on subsys/bluetooth/host/att.c:
```diff
@@ -3088,7 +3106,16 @@ int bt_eatt_connect(struct bt_conn *conn, uint8_t num_channels)
+			return err;
+		}
+
+		created += 5; /* Max channels per request */
```
```suggestion
		created += L2CAP_ECRED_CHAN_MAX; /* Max channels per request */
```

hermabe replied:
Currently the macro is local to `l2cap.c`, but can move it to the internal header (and give it a non-misleading name).

Thalley replied:
Please do :)

Thalley commented on include/bluetooth/att.h:
```diff
@@ -59,12 +59,53 @@ extern "C" {
+/** @brief ATT callback structure. */
+struct bt_eatt_cb {
+	/** @brief An ATT channel has been connected. */
+	void (*eatt_chan_connected)(struct bt_att_chan_info *info);
```
```suggestion
	void (*eatt_chan_connected)(const struct bt_att_chan_info *info);
```

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_eatt/prj_autoconnect.conf:
```diff
@@ -0,0 +1,15 @@
+CONFIG_BT_L2CAP_ECRED=y
+CONFIG_BT_EATT_MAX=5
+CONFIG_BT_MAX_CONN=1
+CONFIG_BT_EATT_AUTO_CONNECT=y
```
This is enabled by default, right?

hermabe replied:
Yes, can be removed.

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_eatt/CMakeLists.txt:
```diff
@@ -15,7 +15,7 @@ project(bsim_test_host)
 target_sources(app PRIVATE
   src/main.c
-  src/main_collision.c
+  src/main_connect.c
```
I'm a little confused here. It seems that you are replacing the main_collision test with main_connect. Is that intentional? If so, what about the prj_collision.conf?

hermabe replied:
Just renaming it because the test is not just used for collision anymore.

Thalley replied:
I see, and then you use the `.conf` files to compile different binaries and use that to test different things? 

To me it would seem cleaner to split the test into two, and have a collision test and a auto-connect test.

hermabe replied:
Yes, the same test procedure is used for multiple tests with different configs. The auto-connect one has a slightly different procedure as it does not call `bt_eatt_connect()`. Can move that one out to a separate file.

hermabe replied:
Created two files instead and moved common code to a separate file.

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_eatt/src/main_connect.c:
```diff
@@ -42,6 +42,18 @@ static const struct bt_data ad[] = {
+	       info->mps);
+
+	num_eatt_channels++;
+	if (num_eatt_channels > CONFIG_BT_EATT_MAX) {
+		FAIL("Too many EATT channels connected (%d)\n", num_eatt_channels);
+	}
```
Given that the peripheral also has the same `CONFIG_BT_EATT_MAX`, shouldn't the check have been

```suggestion
	if (num_eatt_channels != CONFIG_BT_EATT_MAX) {
		FAIL("Incorrect EATT channels connected %u/%u\n", num_eatt_channels, CONFIG_BT_EATT_MAX);
	}
```

hermabe replied:
This callback is called for each EATT channel. We can thus not fail the test the first time the callback is called and `num_eatt_channels==1`.

Can update the message to include `CONFIG_BT_EATT_MAX`.

Thalley replied:
I see. Do we fail the test if we don't connect `CONFIG_BT_EATT_MAX` channels somewhere else then?

hermabe replied:
If we connect too few, the test will time out as the condition on line 131 and 198 never becomes false.
```c
while (num_eatt_channels < CONFIG_BT_EATT_MAX) {
```

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_eatt/src/main_connect.c:
```diff
@@ -88,13 +100,18 @@ BT_CONN_CB_DEFINE(conn_callbacks) = {
 		FAIL("Can't enable Bluetooth (err %d)\n", err);
 	}
 
+	struct bt_eatt_cb eatt_cb = {
+		.eatt_chan_connected = eatt_chan_connected,
+	};
```
This should probably have been statically allocated, right?

hermabe replied:
Yes, it has to be alive until shutdown.

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_eatt/src/main_connect.c:
```diff
@@ -104,19 +121,17 @@ static void test_peripheral_main(void)
 		k_sleep(K_MSEC(100));
 	}
 
+#if !defined(CONFIG_BT_EATT_AUTO_CONNECT)
```
You should be able to use `IS_ENABLED` here

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_eatt/src/main_connect.c:
```diff
@@ -104,19 +121,17 @@ static void test_peripheral_main(void)
-	printk("%d enhanced bearers connected\n", num_enhanced);
-	if (num_enhanced != CONFIG_BT_EATT_MAX) {
-		FAIL("Expected %d enhanced bearers, got %d\n", CONFIG_BT_EATT_MAX, num_enhanced);
+	while (num_eatt_channels < CONFIG_BT_EATT_MAX) {
```
```suggestion
	while (num_eatt_channels != CONFIG_BT_EATT_MAX) {
```

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_eatt/src/main_connect.c:
```diff
@@ -152,13 +167,18 @@ static void device_found(const bt_addr_le_t *addr, int8_t rssi, uint8_t type,
 		FAIL("Can't enable Bluetooth (err %d)\n", err);
 	}
 
+	struct bt_eatt_cb eatt_cb = {
+		.eatt_chan_connected = eatt_chan_connected,
+	};
```
Static?

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_eatt/src/main_connect.c:
```diff
@@ -168,18 +188,15 @@ static void test_central_main(void)
 		k_sleep(K_MSEC(100));
 	}
 
+#if !defined(CONFIG_BT_EATT_AUTO_CONNECT)
```
IS_ENABLED

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_eatt/src/main_connect.c:
```diff
@@ -168,18 +188,15 @@ static void test_central_main(void)
-	printk("%d enhanced bearers connected\n", num_enhanced);
-	if (num_enhanced != CONFIG_BT_EATT_MAX) {
-		FAIL("Expected %d enhanced bearers, got %d\n", CONFIG_BT_EATT_MAX, num_enhanced);
+	while (num_eatt_channels < CONFIG_BT_EATT_MAX) {
```
```suggestion
	while (num_eatt_channels != CONFIG_BT_EATT_MAX) {
```

hermabe committed:
Bluetooth: Host: Update connection test to use bt_eatt_count

The test now uses the public bt_eatt_count function to wait until the
expected number of EATT channels are connected.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

hermabe committed:
Bluetooth: Host: Add test for autoconnection of EATT channels

Checks that CONFIG_BT_EATT_MAX EATT channels are created automatically
after connection.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

hermabe committed:
Bluetooth: Host: Move common bsim test code

Much of the test code will be the same for multiple tests, so move it to
a common file.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

hermabe commented on include/bluetooth/att.h:
```diff
@@ -59,12 +59,53 @@ extern "C" {
+
+/** @brief ATT callback structure. */
+struct bt_eatt_cb {
+	/** @brief An ATT channel has been connected. */
```
Yes, there currently is no API to control each channel, so the CID isn't very useful.

The MTU may be useful because the MTU of the channel may be smaller than what is configured, as the peer may not have enough resources.

I am open to adding/removing things from this struct if they are not useful. We can add more later on if needed. Is there other information about the EATT channel that may be useful?

Thalley replied:
Let's keep it as minimum as possible and then expand on it as necessary.

Thalley replied:
Yes :) Then we'll just add what is requested as it goes on.

Thalley commented on subsys/bluetooth/host/att.c:
```diff
@@ -3088,9 +3105,42 @@ int bt_eatt_connect(struct bt_conn *conn, uint8_t num_channels)
+			return err;
+		}
+
+		created += L2CAP_ECRED_CHAN_MAX_PER_REQ;
```
Can `num_channels` be < `L2CAP_ECRED_CHAN_MAX_PER_REQ? 

If so, this is misleading (although wouldn't cause any issue as far as I can tell

hermabe replied:
Yes, `num_channels` can be anything in the range [1, `CONFIG_BT_EATT_MAX`].

`bt_l2cap_ecred_chan_connect()` will request channels `L2CAP_ECRED_CHAN_MAX_PER_REQ` at a time. The last call will request fewer than that if `num_channels % L2CAP_ECRED_CHAN_MAX_PER_REQ != 0`. `created` will then be larger than `i` and exit the loop. It is then not equal to the number of channels requested so the name is somewhat misleading I guess.

Thalley replied:
I suggest the following

1. Have `bt_l2cap_ecred_chan_connect` somehow return the number of channels attempted to connect and use that value to increment
2. Renamed the variable
3. change this line to `created = MIN(created + L2CAP_ECRED_CHAN_MAX_PER_REQ, i);

Thalley commented on subsys/bluetooth/host/att.c:
```diff
@@ -3088,9 +3105,42 @@ int bt_eatt_connect(struct bt_conn *conn, uint8_t num_channels)
+	return 0;
 }
 
+#if defined(CONFIG_BT_EATT) && defined(CONFIG_BT_EATT_AUTO_CONNECT)
```
Does `CONFIG_BT_EATT_AUTO_CONNECT` not depend on `CONFIG_BT_EATT`? If so, this can just be guarded by `CONFIG_BT_EATT_AUTO_CONNECT`, and if not, it probably should be

hermabe replied:
It is hidden behind a `if BT_EATT` in the Kconfig. Should a `depends on BT_EATT` be added to this and the other EATT Kconfig options?

It is not possible to select `CONFIG_BT_EATT_AUTO_CONNECT` without `CONFIG_BT_EATT`, so the `#ifdef` can be simplified.

Thalley replied:
> Should a depends on BT_EATT be added to this and the other EATT Kconfig options?

Not if it is already hidden by a `if BT_EATT` :)

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_eatt/src/common.c:
```diff
@@ -0,0 +1,176 @@
+	}
+}
+
+void central_disconnect(void)
+{
+	while (is_connected) {
+		k_sleep(K_MSEC(100));
+	}
+}
```
The naming of this function does not match the implementation. Is it missing something?

hermabe replied:
I named it in the same way as the peripheral version, which does the actual disconnection. Can rename to `wait_for_disconnect()`.

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_eatt/src/main_autoconnect.c:
```diff
@@ -0,0 +1,52 @@
+	PASS("EATT Central tests Passed\n");
+}
+
+static const struct bst_test_instance test_def[] = { { .test_id = "peripheral_autoconnect",
+						       .test_descr = "Peripheral autoconnect",
+						       .test_post_init_f = test_init,
+						       .test_tick_f = test_tick,
+						       .test_main_f = test_peripheral_main },
+						     { .test_id = "central_autoconnect",
+						       .test_descr = "Central autoconnect",
+						       .test_post_init_f = test_init,
+						       .test_tick_f = test_tick,
+						       .test_main_f = test_central_main },
+						     BSTEST_END_MARKER };
```
I'm seeing this way of indenting structs a lot; is this some auto-clang format stuff? 

```suggestion
static const struct bst_test_instance test_def[] = { 
	{ 
		.test_id = "peripheral_autoconnect",
		.test_descr = "Peripheral autoconnect",
		.test_post_init_f = test_init,
		.test_tick_f = test_tick,
		.test_main_f = test_peripheral_main 
	},
	{ 
		.test_id = "central_autoconnect",
		.test_descr = "Central autoconnect",
		.test_post_init_f = test_init,
		.test_tick_f = test_tick,
		.test_main_f = test_central_main 
	},
	BSTEST_END_MARKER 
};
```

hermabe replied:
Yes, this is the autoformat that chooses a weird format.

hermabe replied:
Adding a trailing comma makes clang-format behave.

Thalley commented on subsys/bluetooth/host/att.c:
```diff
@@ -3088,9 +3105,42 @@ int bt_eatt_connect(struct bt_conn *conn, uint8_t num_channels)
+void eatt_auto_connect(struct bt_conn *conn, uint8_t err)
+{
+	int eatt_err;
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	eatt_err = att_schedule_eatt_connect(conn, CONFIG_BT_EATT_MAX);
+	if (eatt_err < 0) {
+		bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+		BT_WARN("Automatic creation of EATT bearers failed on "
+			"connection %s with error %d",
+			addr, err);
+	}
```
```suggestion

	eatt_err = att_schedule_eatt_connect(conn, CONFIG_BT_EATT_MAX);
	if (eatt_err < 0) {
		BT_WARN("Automatic creation of EATT bearers failed on "
			"connection %s with error %d",
			bt_addr_le_str_real(bt_conn_get_dst(conn)), err);
	}
```

Thalley reviewed (dismissed):
LGTM

hermabe commented:
@jhedberg Any thoughts about the API?

carlescufi commented:
@jhedberg @Vudentz please take a look

jhedberg commented on include/bluetooth/att.h:
```diff
@@ -59,12 +59,52 @@ extern "C" {
 #endif /* CONFIG_BT_EATT */
 
+/* ATT channel information */
+struct bt_att_chan_info {
+	/* The connection the ATT channel belongs to */
+	struct bt_conn *conn;
+};
```
What kind of things do you foresee us extending this struct with?

Why is it called `bt_att_chan_info` instead of `bt_eatt_chan_info`?

hermabe replied:
Initially the callback was also called for the fixed ATT channel, hence the name.

There were more fields in the struct, the MTU, CID and MPS, but they were removed. See https://github.com/zephyrproject-rtos/zephyr/pull/43112#r813145144

hermabe replied:
Bump @jhedberg

jhedberg commented on include/bluetooth/att.h:
```diff
@@ -59,12 +59,52 @@ extern "C" {
+/** @brief ATT callback structure. */
+struct bt_eatt_cb {
+	/** @brief An ATT channel has been connected. */
+	void (*eatt_chan_connected)(const struct bt_att_chan_info *info);
```
Drop the "eatt". It's already in a struct with "eatt" in the name, so repeating this is redundant.

What other kind of callbacks do you foresee us extending this struct with?

hermabe replied:
During development I had a disconnected callback here as well. Now the struct is only here to make it easy to extend if needed in the future. It also seems like the approach used in GATT with `bt_gatt_cb`.

hermabe replied:
Bump @jhedberg

jhedberg commented on tests/bluetooth/bsim_bt/bsim_test_eatt/src/common.c:
```diff
@@ -0,0 +1,175 @@
+};
+
+static bool volatile is_connected;
+int volatile num_eatt_channels;
```
Should be `volatile int`, not `int volatile` as per coding style conventions

Thalley replied:
Didn't know that. Where are this mentioned? 

I guess this also applies to the `bool` just above?

It should probably also be `static`, right?

jhedberg replied:
I'm not sure if it's anywhere explicitly stated, but `volatile <type>` is the common convention, if you check for other places in the tree. It also reads much more naturally if you consider it from an English language perspective ("variable X is a volatile integer").

jhedberg replied:
As for `static` - I think I saw it exported as an extern in some header file

hermabe replied:
- `num_eatt_channels` is checked in the main test procedure, so cannot be `static`.
- Can change to `volatile int`, as it matches west-const used elsewhere.

hermabe commented:
Updated naming and west-volatile.

Vudentz commented on subsys/bluetooth/host/att.c:
```diff
@@ -3088,9 +3105,40 @@ int bt_eatt_connect(struct bt_conn *conn, uint8_t num_channels)
+void eatt_auto_connect(struct bt_conn *conn, uint8_t err)
+{
+	int eatt_err;
+
```
Should this be checking if the connection is actually a central or not? I know you probably want to test if the collision handling works but I don't think it is a good idea to always start connecting like this, also have you tried connecting with something like BlueZ that supports EATT as well and would start connecting immediately as central?

Thalley replied:
This is done by `att_schedule_eatt_connect` which adds a delay for peripherals based on `credit_based_connection_delay`

Vudentz reviewed (dismissed):
The only concern I have is really with connecting EATT immediately after ACL without checking if we are acting as central, as this most likely will force a collision, while the collision can be handled gracefully or not is another thing but if we can avoid the collision in the first place Id say we better have this change so  BT_EATT_AUTO_CONNECT only really automatically connects as a central or have something like a delayed work when acting as peripheral.

hermabe commented:
> The only concern I have is really with connecting EATT immediately after ACL without checking if we are acting as central, as this most likely will force a collision, while the collision can be handled gracefully or not is another thing but if we can avoid the collision in the first place Id say we better have this change so BT_EATT_AUTO_CONNECT only really automatically connects as a central or have something like a delayed work when acting as peripheral.

Initially only the central auto-connected EATT channels, but it was noted that it would be useful for the peripheral as well (https://github.com/zephyrproject-rtos/zephyr/pull/43112#discussion_r813869469).

The peripheral delays the connection request with the same delay that is used by the L2CAP collision mitigation. We may want to change this or improve the collision mitigation in some other way if interoperability becomes a problem.

hermabe commented:
Rebased on main, and a couple small fixes to the auto-connect.

Thalley commented on include/bluetooth/att.h:
```diff
@@ -59,12 +59,51 @@ extern "C" {
+ *
+ * @param conn The connection to send the request on
+ * @param num_channels The number of Enhanced ATT beares to request.
+ * Must be in the range 1 - @kconfig[CONFIG_BT_EATT_MAX], inclusive.
```
Doesn't it need to be `@kconfig{CONFIG_BT_EATT_MAX}`? Curly brackets instead of square?

hermabe commented:
@jhedberg Do you have any more comments on this?

jhedberg commented:
> @jhedberg Do you have any more comments on this?

Yes, I do. Sorry for the delay.

I'd like to get a better understanding of how exactly the API is to be used, and how it is useful. E.g. for the callbacks you have two layers of abstraction, which in its current form only uncovers access to one single piece of information: a `bt_conn` pointer when a channel has been connected. It's not clear to me how this is useful. The only sample code I saw was tracking the number of connected channels with the help of this pointer, but how is that useful in practice, since there doesn't seem to be a way to reference any specific channel that you want to send or receive data over. What about channel creation failures?

hermabe commented:
The plan is to introduce something along the lines of https://github.com/zephyrproject-rtos/zephyr/pull/41909 so that the application can specify whether to send GATT requests/commands/indications/notifications over Unenhanced ATT or EATT. Then it would be useful for the application to know that EATT channels have been connected before trying to send anything over EATT. For example, some profiles require sending notifications over EATT if supported, otherwise indications over ATT shall be used instead.

A callback for connection failure and disconnection may also be useful.

hermabe commented:
> A callback for connection failure and disconnection may also be useful.

@Thalley @jhedberg @alwa-nordic What do you think about adding callbacks for disconnection and connection failures? Anything else that blocks merging this PR?

Thalley commented:
> > A callback for connection failure and disconnection may also be useful.
> 
> @Thalley @jhedberg @alwa-nordic What do you think about adding callbacks for disconnection and connection failures? Anything else that blocks merging this PR?

Can the receiver of those callbacks do anything about the situation? Otherwise we can probably omit them for now.

hermabe commented:
> Can the receiver of those callbacks do anything about the situation? Otherwise we can probably omit them for now.

If connection fails, the application may want to manually retry the EATT connection. The application will then also know to not try to send anything over EATT. Same with the disconnected callback if the peer disconnects. (`bt_eatt_disconnect()` is not currently public, so no way for the zephyr side to disconnect.)

This will be more useful once the API for specifying that an EATT channel shall be used is added. I am fine with merging as is and adding the additional callbacks later if needed.

Thalley commented:
> > Can the receiver of those callbacks do anything about the situation? Otherwise we can probably omit them for now.
> 
> If connection fails, the application may want to manually retry the EATT connection. The application will then also know to not try to send anything over EATT. Same with the disconnected callback if the peer disconnects. (`bt_eatt_disconnect()` is not currently public, so no way for the zephyr side to disconnect.)
> 
> This will be more useful once the API for specifying that an EATT channel shall be used is added. I am fine with merging as is and adding the additional callbacks later if needed.

I think it's better to add when needed. That will also better show the rationale behind the addition

carlescufi commented:
@Vudentz @jhedberg please take another look

jhedberg commented:
> it would be useful for the application to know that EATT channels have been connected before trying to send anything over EATT.

That's fine, but you don't really need a per-channel callback in that case. A single "EATT channel creation complete" callback would be sufficient. It could e.g. have as a parameter the number of EATT channels that were successfully connected.

hermabe commented:
> That's fine, but you don't really need a per-channel callback in that case. A single "EATT channel creation complete" callback would be sufficient. It could e.g. have as a parameter the number of EATT channels that were successfully connected.

That would require some additional housekeeping. `bt_eatt_connect()` will send multiple requests if `num_channels > 5`. The requests may also be retried in case of collision or reject because of insufficient authentication or encryption. A per-request callback is also not so nice for the application, as it shouldn't have to deal with the L2CAP limitation of maximum 5 channels per connection request. See https://github.com/zephyrproject-rtos/zephyr/pull/42268#discussion_r797646226 for some earlier discussion of this.

jhedberg commented:
I'd really prefer to see some real users of this API (not the shell, but actual profiles using it). Do you have any examples of those?

hermabe commented:
> I'd really prefer to see some real users of this API (not the shell, but actual profiles using it). Do you have any examples of those?

For example in HAP/HARC, the profile needs to know if an EATT channel exists to decide if notifications or indications shall be sent.

> 5.5 Hearing Aid Preset Control Point characteristic requirements
If the HARC supports the Preset Control Point characteristic, the Read Preset procedure, and an
enhanced ATT bearer exists between the client and the server, then the HARC shall register for
notifications and indications of this characteristic.
If the HARC supports the Preset Control Point characteristic, the Read Preset procedure, and only the
unenhanced ATT bearer exists between the client and the server, then the HARC shall register for
indications of this characteristic.

jhedberg commented:
> For example in HAP/HARC, the profile needs to know if an EATT channel exists to decide if notifications or indications shall be sent.

Is the code for that available in some tree? That in itself doesn't really fully justify the API in the form that it's now, since the current proposal provides much more than that.

> That would require some additional housekeeping. bt_eatt_connect() will send multiple requests if num_channels > 5. 

Probably not that big of a deal. In addition to a callback of EATT channel creation completion, maybe a Boolean function to query if any EATT channels are connected would also be useful? Alternatively, such a boolean function could even be the only API we have, if it was made blocking for the duration of the channel creation window, i.e. if you call it earlier it blocks until all channels have been connected.

jhedberg commented:
Just wanted to reiterate the main point why I'm not really happy with the current proposal: having a callback for each connected channel doesn't feel right when we have no way to address a specific channel: you don't get any reference to it in the callback, or when receiving data, and there's also no way to select a specific channel when sending data. Because of this a per-channel callback feels a bit pointless.

Thalley commented:
@hermabe @jhedberg 

@hermabe Is correct that in HAP we need to know if we can send a notification on an EATT channel and if not, then we will send an indication on an ATT channel. 

HAP does not necessarily need to know that an EATT channel has connected, but will need some way of doing

```
if (eatt_connected) {
    send notification on EATT channel
} else {
    send indication on ATT channel
}
```

@jhedberg The is no code for this in the tree at the moment. We are in the progress of adding the HAS that needs it, but haven't gotten to that part yet, and for obvious reasons we can't really upstream that before we have an API that supports the use case. 

HAP/HAS thus does not the connected/disconnected callbacks, as long as it can by some other means ensure that notifications are sent on an EATT channel and if that is not possible, then it needs to know that too.

jhedberg commented:
> for obvious reasons we can't really upstream that before we have an API that supports the use case

That's fine - I wasn't asking for upstream code, but just any tree where it can be reviewed and the API usage can be better understood. Please take a look at my concerns and proposals first however (e.g. the option of not having a callback at all)

Thalley commented:
> > for obvious reasons we can't really upstream that before we have an API that supports the use case
> 
> That's fine - I wasn't asking for upstream code, but just any tree where it can be reviewed and the API usage can be better understood. Please take a look at my concerns and proposals first however (e.g. the option of not having a callback at all)

@MariuszSkamra Do we have any code snippets for HAS that we could show here? I don't recall haven seen one. 

I'm OK with not having the callbacks if we do not have a use case for it. I think there are other ways we can give information to upper layers if EATT channels are available.

hermabe commented:
> I think there are other ways we can give information to upper layers if EATT channels are available.

We have `bt_eatt_count()` which was added for use in testing before the connected callback was added. We can make that public.

Thalley commented:
> > I think there are other ways we can give information to upper layers if EATT channels are available.
> 
> We have `bt_eatt_count()` which was added for use in testing before the connected callback was added. We can make that public.

That would work fine for the LE Audio use cases. We also need the option to require something sent on an EATT channel, so if we add that to the e.g. `bt_gatt_notify` function and that returns a specific error code, that is also good enough.

hermabe commented:
>  We also need the option to require something sent on an EATT channel, so if we add that to the e.g. bt_gatt_notify function and that returns a specific error code, that is also good enough.

That is the plan. (https://github.com/zephyrproject-rtos/zephyr/pull/41909)

hermabe committed:
Bluetooth: Host: Make bt_eatt_count public

bt_eatt_count returns the number of EATT channels connected for a given
connection.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

jhedberg commented on include/bluetooth/att.h:
```diff
@@ -59,10 +59,37 @@ extern "C" {
+ *
+ * @param conn The connection to get the number of EATT channels for.
+ *
+ * @return The number of EATT channels connected.
+ * @retval -EINVAL if @p conn is NULL.
+ * @retval -ENOTCONN @p conn is not connected.
+ */
+int bt_eatt_count(struct bt_conn *conn);
```
I'd keep the original `size_t` return value and simply return `0` in the situations where you're now trying to return a negative error value.

jhedberg replied:
On that note, it would then also be nice to be consistent with the `num_channels` parameter to `bt_eatt_connect()`, i.e. I'd probably change that to `size_t` as well.

Thalley replied:
> I'd keep the original `size_t` return value and simply return `0` in the situations where you're now trying to return a negative error value.

Not sure I agree with this, since there can be multiple reasons why it fails. 

One alternative that may be better is to use the `ssize_t` which is commonly used to return -1 in case of error. Then we have values -1 and 0 to work with for error or unsuccessful conditions.

Thalley replied:
We use `ssize_t` in GATT for that actually, but (unrelated to this) we seem to misusing it, as `ssize_t` is only guaranteed to hold `[-1, SSIZE_MAX]`, but we use it with `BT_GATT_ERR` and an ATT error code, which certainly can be less than `-1`...

jhedberg replied:
@Thalley I think we need to look at the actual errors and consider what useful stuff the caller can do with that information. If you’re disconnected, then returning 0 is just the truth (you can t have any channels if you’re not connected). The NULL pointer case we can discuss, but that could even warrant an assert.

Thalley replied:
I see your point.

hermabe commented:
@carlescufi I think this is ready to merge now.
